Allows for dialing of multiple parties.

TACambron opened the pull request:
Added code primarily to the originateChannel method to allow for
multiple hardcoded phones to be dialed upon StasisStart. When one
dialed phone answers, the other dialed phones are hung up. If all
dialed phones hangup or do not answer, the inbound channel is hungup.
Added two more tests for functionality.

Commit:
Allows for dialing of multiple parties.

Added code primarily to the originateChannel method to allow for
multiple hardcoded phones to be dialed upon StasisStart. When one
dialed phone answers, the other dialed phones are hung up. If all
dialed phones hangup or do not answer, the inbound channel is hungup.
Added two more tests for functionality.

samuelg commented on lib/sla.js:
```diff
@@ -61,35 +93,94 @@ function addChannelsToBridge(channel, bridge) {
+  var originate = [];
+  var numToHangup = dialed.length;
+  var station = ['SIP/phone1', 'SIP/phone2'];
+  dialed.forEach(function (channel, i, dialed) {
```
i is not very descriptive. At the very least call it index.

This would be simpler if you stored an array of objects instead that each contain a channel, station, et cetera. You can then just iterate over the objects and use object.channle or object.station instead of having to use an index into an array.

samuelg commented on lib/sla.js:
```diff
@@ -61,35 +93,94 @@ function addChannelsToBridge(channel, bridge) {
+      defer.resolve(line);
+    });
+    inbound.once('ChannelHangupRequest', function(event, line) {
+      defer.resolve(null);
```
You can just call resolve with no argument. No need to explicitly call it with null.

samuelg commented on lib/sla.js:
```diff
@@ -61,35 +93,94 @@ function addChannelsToBridge(channel, bridge) {
+      });
+    });
+    dialed[i].once('ChannelDestroyed', function(event) {
+      numToHangup -= 1;
```
You could also use a filter to remove the channel from the array and check for array.length === 0 instead of keeping track of a separate variable.

samuelg commented:
I only looked at originateChannel. Once Chris's PR has been merged and you have rebased I'll take a second look.

Commit:
One test that is still failing

TACambron committed:
Fixed a previously fixed trunk hangup issue again.

Added a hangup function inside of the dialed channel hangup listener
that hangs up the inbound trunk when all dialed channels hangup. I
have previously fixed this problem and somehow managed to delete the
fix. It has been fixed again.

samuelg commented on tests/sla_tests.js:
```diff
@@ -125,11 +125,14 @@ var getMockChannel = function() {
     channelId += 1;
     this.originate = function(input, cb) {
-      this.dialed = true;
+      this['dialed'] = 'yes';
```
Looks like you overwrote changes made by Chris that have already been merged to master.

samuelg commented on lib/sla.js:
```diff
@@ -102,26 +102,45 @@ function addChannelsToBridge(inbound, dialed, bridge) {
+      });
+    });
+    dialed[index].once('ChannelDestroyed', function(event) {
+      numToHangup -= 1;
```
Why not just remove the channel from the dialed array and check the length instead of having a numToHangup variable to keep track of on top of the array?

samuelg commented on tests/sla_tests.js:
```diff
@@ -140,20 +143,20 @@ var getMockChannel = function() {
     this.hangup = function(cb) {
       var self = this;
-      this.wasHungup = true;
+      this['hungup'] = 'yes';
```
Same here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -140,20 +143,20 @@ var getMockChannel = function() {
           channels = channels.filter(function(channel) {
             return channel !== self;
           });
-          cb(null);
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -140,20 +143,20 @@ var getMockChannel = function() {
     };
     this.answer = function(cb) {
-      this.wasAnswered = true;
+      this['answered'] = 'yes';
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -221,8 +224,8 @@ describe('SLA Bridge and Channels Tester', function() {
       setTimeout(function() {
-        if(dialed[0] && dialed[0].wasAnswered && bridgeChannels.length !== 0 &&
-          isMixing && inbound.wasAnswered) {
+        if(dialed[0] && dialed[0]['answered'] && bridgeChannels.length !== 0 &&
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -243,8 +246,8 @@ describe('SLA Bridge and Channels Tester', function() {
       setTimeout(function() {
-        if(dialed[0] && !dialed[0].wasAnswered && bridgeChannels.length === 0 &&
-          isMixing && inbound.wasAnswered) {
+        if(dialed[0] && !dialed[0]['answered'] && bridgeChannels.length === 0 &&
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -257,7 +260,7 @@ describe('SLA Bridge and Channels Tester', function() {
     var client = getMockClient();
     var inbound = getMockChannel();
-    inbound.inbound = true;
+    inbound['inbound'] = 'yes';
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -265,8 +268,8 @@ describe('SLA Bridge and Channels Tester', function() {
       setTimeout(function() {
-        if(bridges.length === 0 && !dialed[0] && inbound.inbound &&
-          !inbound.wasAnswered && inbound.wasHungup && channels.length === 0) {
+        if(bridges.length === 0 && !dialed[0] && inbound['inbound'] &&
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -279,7 +282,7 @@ describe('SLA Bridge and Channels Tester', function() {
     var client = getMockClient();
     var inbound = getMockChannel();
-    inbound.inbound = true;
+    inbound['inbound'] = 'yes';
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -290,8 +293,8 @@ describe('SLA Bridge and Channels Tester', function() {
     function earlyHangup() {
       setTimeout(function() {
         if(isMixing && channels.length === 0 && bridges.length === 1 &&
-          inbound.inbound && inbound.wasAnswered && inbound.wasHungup &&
```
And here.

samuelg commented on tests/sla_tests.js:
```diff
@@ -299,4 +302,58 @@ describe('SLA Bridge and Channels Tester', function() {
+    cancelDialing();
+    function cancelDialing() {
+      setTimeout(function() {
+        if(isMixing && channels.length === 2 && bridges.length === 1 &&
```
You'll have to match isAnswered instead of ['answered'].

samuelg commented on tests/sla_tests.js:
```diff
@@ -299,4 +302,58 @@ describe('SLA Bridge and Channels Tester', function() {
+    }, asyncDelay);
+    function failToAnswer() {
+      setTimeout(function() {
+        if(isMixing && channels.length === 0 && bridges.length === 1 &&
```
You'll have to match isAnswered instead of ['answered'].

TACambron committed:
Fixed previously corrected findings.

Managed to revert some corrections that I had already made, so I
fixed those once again. I also removed the numToHangup variable.

TACambron committed:
Updated answered and hungup properties to read wasAnswered and wasHungup.

Updated the nomenclature to fit in with previous naming schemes.

samuelg commented:
Looks good. Make sure you squash your commits into 1 commit before merging.

TACambron commented:
Squashed into #11
